### ISSUE TITLE
linter: consistent-type-definitions rule does not skip over comments and will highlight comments instead of keywords

### ISSUE THREAD

**Issue by bradzacher** [oxc-project/oxc#14833]
### What version of Oxlint are you using?

playground

### What command did you run?

https://playground.oxc.rs/#eNptUk2P00AM/SuVj6uuFpCQUBEnlkor8SWEuBAOsxMnzDIZj2yn7SrKf8eTNlMOnJJ5/np+zxN42EGLPjrGzd3NRp8zbm7uzt/vm3ebaX7bpCbBFgh2E/CYyieGpLBTHnELHfHgVJFh17ko/yIPV0zZJSmBigSh6BTb+2W600BJavD4OyhKdh4rNLjUx+vT05AZ5VoinrKFz6TkeXikuL5811/S5i1kx1LIToAnxSQ21zRQOdmOLkY6fkMdOX0ZVUKL+zH5Qm1tVWYiH/CrCVb4ntGl7sebh6QcrKOvAUGjrcF/YCa+oMah6Hfm4Cl1wdjB1ECOY2/lDex+NlAcEM8hawO/tg1406onDljiluyJGb0mk8CABqjrGpgtkcd4zTEyYkvqbWl322IXUliUXmqOjpMVzVZ3NejMSx33aB4DyqsXL1+bNqPgfanHPfH76ET2AWNbN8VTRg6DzXLRLCVztG5s0SFoRT+hutapq3L8/xQmEJMzm6jIyZmZq4OeWuxxOcRUTq36/CQt2UFfPEmJdOm1IhH7NXfpsh7QZK/1umxzyh/xgLXpH8T82Q24XtpSakZT3JvppeKA/EiynmbZJz2ZNSVU/2aDF/mXGfVvLniRed1p/gsblUdw

### What does your `.oxlintrc.json` config file look like?

`consistent-type-definitions: 'error'`

### Input:

```ts
declare /* type */ type T = {};
```

### Expected:

```ts
declare /* type */ type T = {};
                   ^^^^ error here
```

### Actual:

```ts
declare /* type */ type T = {};
           ^^^^ error here
```

Because oxc/oxlint doesn't work with tokens it is unable to distinguish between comment text and code text.

---

Side note that I suspect that if I had a bit of a poke around I could find a few bugs like this. It's the major weaknesses of relying on `String::find` on top of source code instead of having a CST to inspect.

**Comment by MichaReiser** [comment]
Something we're doing in Ruff, which works fine and doesn't require a whole CST, is that we store the tokens alongside the parsed AST, and we can poke into the tokens when the information in the AST isn't enough. 

We also have a simplified lexer that can lex most tokens except strings. It allows us to do ad-hoc lexing when needed (e.g. what's the first token after offset XX, but skipping over comments. 

A CST would obviously be nice and has some other benefits, e.g. that it enables edits that preserve source formatting and comments.

**Comment by bradzacher** [comment]
Yeah eslint doesn't use a full CST - just the tokens. It was why I began poking around here.

I've had a few people at work ask when we could look into migrating our internal eslint rules to oxlint and I knew there was a dozen or so that wouldn't work due to not having the tokens.

So I was curious how they had worked around the lack of tokens when they migrated some rules from typescript-eslint and that's when I found this bug.

The fixer for the rule also does lose comments in some locations due to the fact that the fixer replaces the entire node's range with the new type. For comparison the typescript-eslint fixer picks individual tokens to change or delete to avoid mutating the source code too much to help preserve comments.

Tokens would be a boon to making rules safer and more resilient to such bugs.

**Comment by overlookmotel** [comment]
I can see access to tokens or a CST would be useful, but our rationale is this:

Lint errors should be rare. If you get errors/warnings in a lint run, you'll tend to fix them (or disable them with an `// oxlint-disable` comment). So on the next run, no errors.

Therefore, the hot path in linter is "no errors in this file". Storing tokens, or constructing a CST, when none of that data will be used for the vast majority of files being linted would likely be a large drag on performance.

However, conversely, once you *do* hit an error, it's no problem to do a bunch of work - because that expensive branch is rarely taken. So we could look at providing better APIs to lazily tokenize (or similar), in order to make it easier to write accurate diagnostic/fix code.

**Comment by bradzacher** [comment]
For sure - the rationale makes sense.

But definitely there must be APIs that make it trivial to grab the tokens so that rule authors reach for that instead of doing things like reaching for substring matches cos it's easier.

Substring matches against code will **always** lead to buggy rule reports and even worse - to buggy fixers which produce broken code.

An incorrect report location isn't terrible - it's a minor quibble a user can mostly ignore.
But a broken fixer is lethal - it means a user can't trust autofix-on-save or `--fix` runs because they're scared their code will get broken.

From my experience working on linters - losing that trust is really hard to get back. Cos once you've broken someone's code once they can never ever be 100% sure that you won't do it again.

As an example at FB for a long time they purposely did two things:
1) forcibly disabled autofix-on-save for everyone
2) made their linter CLI show the user each fix and force them to accept it manually

They did this because they had years (in the early days of linters) where so many fixers were outright broken.
It took a lot of effort to rebuild that trust and even then I was only able to enable autofix-on-save with a small allowlist of rules.

**Comment by bradzacher** [comment]
I'd love to start exploring using oxlint in Canva's codebase -- but we won't be able to do it without the token APIs for our internal lint rules.

Would you like a separate proposal to track the creation of a new feature that allows you to access tokens for a node ad-hoc only when there is an active report? I.e. something akin to this (in the rust API): `context.report(|foo| {foo.get_token(node); /* ... */})`

**Comment by camc314** [comment]
> Would you like a separate proposal to track the creation of a new feature that allows you to access tokens for a node ad-hoc only when there is an active report? I.e. something akin to this (in the rust API): context.report(|foo| {foo.get_token(node); /* ... */})

We've introduced `find_next_token_from` for use in the rust rules, this skips comments and searches for certain token, which has covered most of our use cases. Is that enough for you?

**Comment by overlookmotel** [comment]
I know you're keen to port rules to Rust ([Discord thread](https://discord.com/channels/1079625926024900739/1450786398733992016)), but just to point out that all the tokens APIs are now supported in JS plugins.

**Comment by bradzacher** [comment]
Ideally the rust side would have an API that could get you all the tokens for any required node so that you can do sub-range reporting easily without reaching for fuzzy string matching.

EG in this specific issue you could do something like:
```rust
ctx.diagnostic_with_tokens(|tokens, reporter| {
  let type_token = tokens
    .get(decl, TokenKind::Keyword("type"))
    .unwrap();
  reporter.diagnostic_with_fix(
    consistent_type_definitions_diagnostic(
      "interface",
      "type",
      type_token.span,
    ),
    |fix| { ... },
  )
});
```

So this way the usage of tokens is physically limited to the "sad path" (the path where a rule must report) and the happy path cannot do the slow thing.

This would allow for well-defined lint rules that don't rely on syntactic ambiguity of string matching at all. My main concern is that right now with no true tokens API on the rust side there's a lot of want to just reach for fuzzy string matching to do fixing and sub-node-range reporting (eg reporting on just the `type` keyword). But that leaves the door open for heaps of bugs in rules cos code doesn't fit nicely into a fuzzy string matching box.

A first-class tokens API for reporting+fixing would help ensure that all rust rules first reach for sound and non-ambiguous logic first to help avoid such bugs as this one in future. Which would ultimately provide a better foundation for OXLint's growth.

**Comment by overlookmotel** [comment]
We are actually working at the moment on adding an option to Oxc parser to produce a `Vec<Token>` alongside AST, because it's needed for JS plugins. At present to support tokens APIs on JS side we re-parse with TypeScript parser and convert TypeScript's AST to tokens (same as TS-ESLint does). But obviously this is pretty slow! So we're looking to Rustify it.

Once we've implemented that, APIs of the sort you've described will be much more feasible.

**Comment by camc314** [comment]
Thanks for all the context everyone!

I'm looking to close this one out; The original but here has been fixed (https://github.com/oxc-project/oxc/pull/14909).

We've also introduced several new APIs on semantic for searching for tokens such as: `find_next_token_from`, `find_prev_token_from`, and bounded variants like `find_next_token_within` / `find_prev_token_within`. These should cover the common cases where rules need to locate a keyword or punctuation token for diagnostics/fixes, without relying on plain source text substring matches.

We've also shipped rustafied ESTree style tokens alongside the AST for use with JS plugins. This should allow plugin authors to use the same token logic.

I think it's worth considering moving to use these `Vec<Token>` at some point, however I think the new semantic APIs work well enough in the interim.

Thanks all!
